tests/services/lab_test.py

```diff
@@ -0,0 +1,289 @@
+    user_map: UserMap,
+) -> None:
+    lab = obj_factory.labspecs[0]
+    assert user_context.user is not None
+    user = user_context.user
+    username = user.username
+    namespace = user_context.namespace
+    token = user_context.token
+    manager_namespace = config.runtime.namespace_prefix
+    instance_url = config.runtime.instance_url
+    lab_config = config.lab
+    lm = LabManager(
+        username=username,
+        namespace=namespace,
+        manager_namespace=manager_namespace,
+        instance_url=instance_url,
+        user_map=user_map,
+        lab=lab,
+        prepuller_manager=prepuller_manager,
+        logger=logger,
+        lab_config=lab_config,
+        k8s_client=k8s_storage_client,
+        user=user,
+        token=token,
+    )
+    present = await lm.check_for_user()
+    assert present is True  # It should already be in the user map
+    dlm = DeleteLabManager(
+        user_map=user_map, k8s_client=k8s_storage_client, logger=logger
+    )
+    await dlm.delete_lab_environment(username=username)
+    present = await lm.check_for_user()
+    assert present is False  # And now it should not be
+    await lm.create_lab()
+    present = await lm.check_for_user()
+    assert present is True  # And should now have returned
+
+
+@pytest.mark.asyncio
+async def test_get_active_users(
+    obj_factory: TestObjectFactory,
+    user_context: Context,
+) -> None:
+    users = user_context.user_map.running
+    assert len(users) == 1
+    assert users[0] == "wrench"
+
+
+@pytest.mark.asyncio
+async def test_nss(
+    obj_factory: TestObjectFactory,
+    prepuller_manager: PrepullerManager,
+    user_context: Context,
+    logger: BoundLogger,
+    config: Configuration,
+    k8s_storage_client: K8sStorageClient,
+    user_map: UserMap,
+) -> None:
+    lab = obj_factory.labspecs[0]
+    assert user_context.user is not None
+    user = user_context.user
+    username = user.username
+    namespace = user_context.namespace
+    token = user_context.token
+    manager_namespace = config.runtime.namespace_prefix
+    instance_url = config.runtime.instance_url
+    lab_config = config.lab
+    lm = LabManager(
+        username=username,
+        namespace=namespace,
+        manager_namespace=manager_namespace,
+        instance_url=instance_url,
+        user_map=user_map,
+        lab=lab,
+        prepuller_manager=prepuller_manager,
+        logger=logger,
+        lab_config=lab_config,
+        k8s_client=k8s_storage_client,
+        user=user,
+        token=token,
+    )
+    nss = await lm.build_nss()
+    assert nss["/etc/passwd"].endswith(
+        "lsst_lcl:x:1000:1000::/home/lsst_lcl:/bin/bash\n"
+        "rachel:x:1101:1101:Rachel (?):/home/rachel:/bin/bash\n"
+    )
+    assert nss["/etc/group"].endswith(
+        "input:x:999:\n"
+        "rachel:x:1101:\n"
+        "lunatics:x:2028:rachel\n"
+        "mechanics:x:2001:rachel\n"
+        "storytellers:x:2021:rachel\n"
+    )
```

Comment:
This test has fallen victim to a problem that comes up a lot in pytest: it's not at all obvious why you're probing for the results that you're probing for because all the data is hiding in fixtures and then is repeated here. This makes the test hard to understand, and it also makes the test fragile because if you change the fixture (even in unrelated ways) the test will break.

What I've found is better is to go either one way or the other: either stop using fixtures for all of this stuff and explicitly pass data into the `LabManager` that's present as literal strings in this test, and then check the results for the same strings. Or parameterize your asserts so that rather than hard-coding the expected results, you also look at the fixture and interpolate (with f-strings) the expected values. Either one works, although I've found I prefer the former because it makes the tests more self-contained and easier to maintain.